source-batchsql/main.go

```diff
@@ -0,0 +1,480 @@
+	var endpointSchema, err = schemagen.GenerateSchema("Batch SQL", &Config{}).MarshalJSON()
+	if err != nil {
+		fmt.Println(fmt.Errorf("generating endpoint schema: %w", err))
+	}
+	resourceSchema, err := schemagen.GenerateSchema("Batch SQL Resource Spec", &Resource{}).MarshalJSON()
+	if err != nil {
+		return nil, fmt.Errorf("generating resource schema: %w", err)
+	}
+
+	return &pc.Response_Spec{
+		ConfigSchemaJson:         json.RawMessage(endpointSchema),
+		ResourceConfigSchemaJson: json.RawMessage(resourceSchema),
+		DocumentationUrl:         "https://go.estuary.dev/source-batchsql",
+	}, nil
+}
+
+func (driver) Apply(ctx context.Context, req *pc.Request_Apply) (*pc.Response_Applied, error) {
+	return &pc.Response_Applied{ActionDescription: ""}, nil
+}
+
+const discoverQuery = `
+SELECT table_schema, table_name, table_type
+  FROM information_schema.tables
+  WHERE table_schema NOT IN ('pg_catalog', 'information_schema', 'pg_internal', 'catalog_history', 'cron');
+`
+
+func (driver) Discover(ctx context.Context, req *pc.Request_Discover) (*pc.Response_Discovered, error) {
+	var cfg Config
+	if err := pf.UnmarshalStrict(req.ConfigJson, &cfg); err != nil {
+		return nil, fmt.Errorf("parsing endpoint config: %w", err)
+	}
+
+	// Validate connection to database
+	log.WithFields(log.Fields{
+		"address":  cfg.Address,
+		"user":     cfg.User,
+		"database": cfg.Database,
+	}).Info("connecting to database")
+	var db, err = sql.Open("pgx", cfg.ToURI())
+	if err != nil {
+		return nil, fmt.Errorf("error opening database connection: %w", err)
+	} else if err := db.PingContext(ctx); err != nil {
+		return nil, fmt.Errorf("error pinging database: %w", err)
+	}
+	defer db.Close()
+
+	rows, err := db.QueryContext(ctx, discoverQuery)
+	if err != nil {
+		return nil, fmt.Errorf("error executing discovery query: %w", err)
+	}
+	defer rows.Close()
+
+	var bindings []*pc.Response_Discovered_Binding
+	for rows.Next() {
+		var tableSchema, tableName, tableType string
+		if err := rows.Scan(&tableSchema, &tableName, &tableType); err != nil {
+			return nil, fmt.Errorf("error scanning result row: %w", err)
+		}
+
+		var resourceName = recommendedCatalogName(tableSchema, tableName)
+		var res = &Resource{
+			Name:     resourceName,
+			Template: fmt.Sprintf(`SELECT * FROM "%s"."%s";`, tableSchema, tableName),
+		}
+		var resourceConfigJSON, err = json.Marshal(res)
+		if err != nil {
+			return nil, fmt.Errorf("error serializing resource spec: %w", err)
+		}
+
+		bindings = append(bindings, &pc.Response_Discovered_Binding{
+			RecommendedName:    pf.Collection(resourceName),
+			ResourceConfigJson: resourceConfigJSON,
+			DocumentSchemaJson: minimalSchema,
+			Key:                defaultKey,
+		})
+	}
+
+	return &pc.Response_Discovered{Bindings: bindings}, nil
+}
+
+var catalogNameSanitizerRe = regexp.MustCompile(`(?i)[^a-z0-9\-_.]`)
+
+func recommendedCatalogName(schema, table string) string {
```

Comment:
This function is copied from `sqlcapture` and could be exported from there in a real implementation of this concept.